Support batch insert for pooled sqlite connections and/or make it clear that connection pooling shouldn't be used with SQLite

This was used to help reproduce the issue.

https://gitlab.com/tahahawa/discord-markov-bot/blob/eae22b1744348b70b5ceccbe9db148b34352b41b/src/commands/helper.rs

The code: 
```Rust
let conn = sql_pool.get().unwrap();
let _ = diesel::replace_into(messages::table)
    .values(message_vec)
    .execute(&conn)
    .expect("Error inserting values");
```

Causes:

```
error[E0277]: the trait bound `diesel::sqlite::Sqlite: diesel::backend::SupportsDefaultKeyword` is not satisfied
   --> src/commands/helper.rs:184:10
    |
184 |         .execute(&conn)
    |          ^^^^^^^ the trait `diesel::backend::SupportsDefaultKeyword` is not implemented for `diesel::sqlite::Sqlite`
    |
    = note: required because of the requirements on the impl of `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` for `diesel::insertable::BatchInsert<'_, models::InsertableMessage, schema::messages::table>`
    = note: required because of the requirements on the impl of `diesel::query_builder::QueryFragment<diesel::sqlite::Sqlite>` for `diesel::query_builder::InsertStatement<schema::messages::table, diesel::insertable::BatchInsert<'_, models::InsertableMessage, schema::messages::table>, diesel::query_builder::insert_statement::Replace>`
    = note: required because of the requirements on the impl of `diesel::query_dsl::load_dsl::ExecuteDsl<diesel::r2d2::PooledConnection<diesel::r2d2::ConnectionManager<diesel::SqliteConnection>>, diesel::sqlite::Sqlite>` for `diesel::query_builder::InsertStatement<schema::messages::table, diesel::insertable::BatchInsert<'_, models::InsertableMessage, schema::messages::table>, diesel::query_builder::insert_statement::Replace>`
```

Current fix:
Change `&conn` to `&*conn`

## Setup

### Versions

- **Rust:**  1.28.0
- **Diesel:**  1.3.2
- **Database:** SQLite with R2D2 pooling
- **Operating System**  Arch Linux

### Feature Flags

- **diesel:**  ["sqlite","r2d2"]

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) for similar issues.
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)

--- a/diesel/src/query_builder/insert_statement/mod.rs
+++ b/diesel/src/query_builder/insert_statement/mod.rs
@@ -23,7 +23,9 @@ use crate::query_dsl::RunQueryDsl;
 use crate::query_source::{Column, Table};
 use crate::result::QueryResult;
 #[cfg(feature = "sqlite")]
-use crate::sqlite::{Sqlite, SqliteConnection};
+use crate::sqlite::Sqlite;
+#[cfg(feature = "sqlite")]
+use crate::Connection;
 
 /// The structure returned by [`insert_into`].
 ///
@@ -202,15 +204,15 @@ where
 }
 
 #[cfg(feature = "sqlite")]
-impl<'a, T, U, Op> ExecuteDsl<SqliteConnection> for InsertStatement<T, BatchInsert<'a, U, T>, Op>
+impl<'a, T, U, Op, C> ExecuteDsl<C, Sqlite> for InsertStatement<T, BatchInsert<'a, U, T>, Op>
 where
+    C: Connection<Backend = Sqlite>,
     &'a U: Insertable<T>,
     InsertStatement<T, <&'a U as Insertable<T>>::Values, Op>: QueryFragment<Sqlite>,
     T: Copy,
     Op: Copy,
 {
-    fn execute(query: Self, conn: &SqliteConnection) -> QueryResult<usize> {
-        use crate::connection::Connection;
+    fn execute(query: Self, conn: &C) -> QueryResult<usize> {
         conn.transaction(|| {
             let mut result = 0;
             for record in query.records.records {
@@ -284,15 +286,15 @@ where
 }
 
 #[cfg(feature = "sqlite")]
-impl<T, U, Op> ExecuteDsl<SqliteConnection>
+impl<T, U, Op, C> ExecuteDsl<C, Sqlite>
     for InsertStatement<T, OwnedBatchInsert<ValuesClause<U, T>, T>, Op>
 where
+    C: Connection<Backend = Sqlite>,
     InsertStatement<T, ValuesClause<U, T>, Op>: QueryFragment<Sqlite>,
     T: Copy,
     Op: Copy,
 {
-    fn execute(query: Self, conn: &SqliteConnection) -> QueryResult<usize> {
-        use crate::connection::Connection;
+    fn execute(query: Self, conn: &C) -> QueryResult<usize> {
         conn.transaction(|| {
             let mut result = 0;
             for value in query.records.values {

--- a/diesel/src/r2d2.rs
+++ b/diesel/src/r2d2.rs
@@ -212,6 +212,29 @@ where
     }
 }
 
+impl<M> crate::migration::MigrationConnection for PooledConnection<M>
+where
+    M: ManageConnection,
+    M::Connection: crate::migration::MigrationConnection,
+    Self: Connection,
+{
+    fn setup(&self) -> QueryResult<usize> {
+        (&**self).setup()
+    }
+}
+
+impl<Changes, Output, M> crate::query_dsl::UpdateAndFetchResults<Changes, Output>
+    for PooledConnection<M>
+where
+    M: ManageConnection,
+    M::Connection: crate::query_dsl::UpdateAndFetchResults<Changes, Output>,
+    Self: Connection,
+{
+    fn update_and_fetch(&self, changeset: Changes) -> QueryResult<Output> {
+        (&**self).update_and_fetch(changeset)
+    }
+}
+
 #[cfg(test)]
 mod tests {
     use std::sync::mpsc;

--- a/diesel/src/r2d2.rs
+++ b/diesel/src/r2d2.rs
@@ -16,8 +16,7 @@ use std::convert::Into;
 use std::fmt;
 use std::marker::PhantomData;
 
-use crate::backend::UsesAnsiSavepointSyntax;
-use crate::connection::{AnsiTransactionManager, SimpleConnection};
+use crate::connection::{SimpleConnection, TransactionManager};
 use crate::deserialize::{Queryable, QueryableByName};
 use crate::prelude::*;
 use crate::query_builder::{AsQuery, QueryFragment, QueryId};
@@ -128,12 +127,10 @@ where
 impl<M> Connection for PooledConnection<M>
 where
     M: ManageConnection,
-    M::Connection:
-        Connection<TransactionManager = AnsiTransactionManager> + R2D2Connection + Send + 'static,
-    <M::Connection as Connection>::Backend: UsesAnsiSavepointSyntax,
+    M::Connection: Connection + R2D2Connection + Send + 'static,
 {
     type Backend = <M::Connection as Connection>::Backend;
-    type TransactionManager = <M::Connection as Connection>::TransactionManager;
+    type TransactionManager = PooledConnectionTransactionManager<M>;
 
     fn establish(_: &str) -> ConnectionResult<Self> {
         Err(ConnectionError::BadConnection(String::from(
@@ -171,7 +168,47 @@ where
     }
 
     fn transaction_manager(&self) -> &Self::TransactionManager {
-        (&**self).transaction_manager()
+        // This is actually fine because we have an #[repr(transparent)]
+        // on LoggingTransactionManager, which means the layout is the same
+        // as the inner type
+        // See the ref-cast crate for a longer version: https://github.com/dtolnay/ref-cast
+        unsafe {
+            &*((&**self).transaction_manager() as *const _ as *const Self::TransactionManager)
+        }
+    }
+}
+
+#[doc(hidden)]
+#[repr(transparent)]
+#[allow(missing_debug_implementations)]
+pub struct PooledConnectionTransactionManager<M>
+where
+    M: ManageConnection,
+    M::Connection: Connection,
+{
+    inner: <M::Connection as Connection>::TransactionManager,
+}
+
+impl<M> TransactionManager<PooledConnection<M>> for PooledConnectionTransactionManager<M>
+where
+    M: ManageConnection,
+    PooledConnection<M>: Connection,
+    M::Connection: Connection,
+{
+    fn begin_transaction(&self, conn: &PooledConnection<M>) -> QueryResult<()> {
+        self.inner.begin_transaction(&**conn)
+    }
+
+    fn rollback_transaction(&self, conn: &PooledConnection<M>) -> QueryResult<()> {
+        self.inner.rollback_transaction(&**conn)
+    }
+
+    fn commit_transaction(&self, conn: &PooledConnection<M>) -> QueryResult<()> {
+        self.inner.commit_transaction(&**conn)
+    }
+
+    fn get_transaction_depth(&self) -> u32 {
+        self.inner.get_transaction_depth()
     }
 }
 

